File: app/src/main/java/ru/samtakoy/core/model/utils/cbuild/OneCardParser.java
package ru.samtakoy.core.model.utils.cbuild;

import java.util.HashMap;

import ru.samtakoy.core.model.Tag;

public class OneCardParser {




    enum BuildPhase{NONE, QUESTION, ANSWER, FINISHED};

    private HashMap<String, Tag> mTagMap;

    private StringBuilder mText;
    private CardBuilder mCurCardBuilder;

    private boolean mIsTextClosed;

    public OneCardParser(HashMap<String, Tag>tagMap) {
        mTagMap = tagMap;
        mText = new StringBuilder();
        mIsTextClosed = false;
    }

    public void reset(){

        mCurCardBuilder = null;
        mText.setLength(0);
        mIsTextClosed = false;
    }

    public void openQuestion(){
        reset();
        mCurCardBuilder = new CardBuilder();
    }

    public void openAnswer(){

        mCurCardBuilder.setQuestion(mText.toString());
        mText.setLength(0);
    }

    public void addTagsLine(String tagsLine){
        if(tagsLine.trim().length()>1){
            String[] tags = tagsLine.split("#");
            for(String tagTitle:tags){
                // первая будет пустая
                if(tagTitle.length() > 0){
                    addTag(tagTitle.trim());
                }
            }
        }

        mIsTextClosed = true;
    }

    private void addTag(String tagTitle){
        String key = Tag.titleToKey(tagTitle);
        Tag tag = mTagMap.get(key);
        if(tag == null){
            tag = Tag.createNew(tagTitle);
            mTagMap.put(tag.getMapKey(), tag);
        }
        mCurCardBuilder.addTag(tag);
    }

    public void addImage(String imageLine) {

        imageLine = imageLine.trim();
        if(imageLine.length()>1){
            String[] images = imageLine.split(",");
            for(String img:images){
                img = img.trim();
                if(img.length() > 0){
                    mCurCardBuilder.addImage(img);
                }
            }
        }
    }

    public void addLine(String line, boolean isFirstLine){
        if(!mIsTextClosed) {
            if (!isFirstLine) {
                mText.append(CBuilderConst.LINE_BREAK);
            }
            mText.append(line);
        }
    }

    public CardBuilder getBuilder(Long qPackId){

        if(!isCardOpened()){
            return null;
        }

        // ?
        //mCurCardBuilder.setQPackId(qPackId);
        mCurCardBuilder.setAnswer(mText.toString());

        CardBuilder result = mCurCardBuilder;

        reset();
        return  result;
    }

    public boolean isCardOpened(){
        return  mCurCardBuilder != null;
    }

    public boolean isValidCard() {
        return  mCurCardBuilder != null && mCurCardBuilder.isValid();
    }

    public void markCardToRemove(Long cardId) {
        setCardId(cardId);
        mCurCardBuilder.toRemove();
    }

    public void setCardId(Long cardId) {
        mCurCardBuilder.setId(cardId);
    }

}

File: module/domain/import/api/src/main/kotlin/ru/samtakoy/domain/importcards/batch/utils/builder/OneCardParser.kt
package ru.samtakoy.domain.importcards.batch.utils.builder

import ru.samtakoy.domain.cardtag.ConcurrentTagMap
import ru.samtakoy.domain.cardtag.Tag
import ru.samtakoy.domain.cardtag.tagTitleToKey

class OneCardParser(
    private val mTagMap: ConcurrentTagMap
) {

    private val mText: StringBuilder
    private var mCurCardBuilder: CardBuilder? = null

    private var mIsTextClosed = false

    init {
        mText = StringBuilder()
    }

    fun reset() {
        mCurCardBuilder = null
        mText.setLength(0)
        mIsTextClosed = false
    }

    fun openQuestion() {
        reset()
        mCurCardBuilder = CardBuilder()
    }

    fun openAnswer() {
        mCurCardBuilder!!.setQuestion(mText.toString())
        mText.setLength(0)
    }

    suspend fun addTagsLine(tagsLine: String) {
        if (tagsLine.trim { it.isWhitespace() }.length > 1) {
            tagsLine
                .split("#".toRegex())
                .asSequence()
                .map { it.trim { it.isWhitespace() } }
                .filter { it.isNotEmpty() }
                .forEach { tagTitle ->
                    // первая будет пустая
                    if (tagTitle.isNotEmpty()) {
                        addTag(tagTitle.trim { it.isWhitespace() })
                    }
                }
        }

        mIsTextClosed = true
    }

    private suspend fun addTag(tagTitle: String) {
        val key: String = tagTitle.tagTitleToKey()
        mTagMap.addIfNone(key) {
            Tag(id = 0L, title = tagTitle)
        }
        mCurCardBuilder!!.addTagKey(key)
    }

    fun addImage(imageLine: String) {
        var imageLine = imageLine
        imageLine = imageLine.trim { it <= ' ' }
        if (imageLine.length > 1) {
            val images = imageLine.split(",".toRegex()).dropLastWhile { it.isEmpty() }.toTypedArray()
            for (img in images) {
                var img = img
                img = img.trim { it <= ' ' }
                if (img.length > 0) {
                    mCurCardBuilder!!.addImage(img)
                }
            }
        }
    }

    fun addLine(line: String?, isFirstLine: Boolean) {
        if (!mIsTextClosed) {
            if (!isFirstLine) {
                mText.append(CBuilderConst.LINE_BREAK)
            }
            mText.append(line)
        }
    }

    fun getBuilder(qPackId: Long?): CardBuilder? {
        if (!this.isCardOpened) {
            return null
        }

        // ?
        // mCurCardBuilder!!.qPackId = qPackId ?: CBuilderConst.NO_ID
        mCurCardBuilder!!.setAnswer(mText.toString())

        val result = mCurCardBuilder

        reset()
        return result
    }

    val isCardOpened: Boolean
        get() = mCurCardBuilder != null

    val isValidCard: Boolean
        get() = mCurCardBuilder != null && mCurCardBuilder!!.isValid

    fun markCardToRemove(cardId: Long) {
        setCardId(cardId)
        mCurCardBuilder!!.toRemove()
    }

    fun setCardId(cardId: Long) {
        mCurCardBuilder!!.cardId = cardId
    }
}